src/ngraph/runtime/cpu/pass/cpu_collapse_dims.cpp

```diff
@@ -0,0 +1,155 @@
+/*******************************************************************************
+* Copyright 2017-2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "cpu_collapse_dims.hpp"
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include <unordered_set>
+#include "ngraph/graph_util.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/broadcast.hpp"
+#include "ngraph/op/reshape.hpp"
+
+using namespace ngraph;
+
+static void get_default_order(std::vector<size_t>& order, size_t rank)
+{
+    for (size_t i = 0; i < rank; i++)
+    {
+        order.push_back(i);
+    }
+}
+
+struct CollapsedDims
+{
+    std::vector<size_t> output_shape;
+    std::vector<bool> is_operated_axis;
+    std::vector<size_t> axis_set;
+    std::vector<size_t> input_shape;
+};
+
+// Fold and collapse axes of output_shape.
+// Contiguous axes that are not being operated on can be collapsed.
+// Contiguous axes that are being operated on are collapsed optionally.
+// Skip size 1 dimensions.
+static void collapse_dims(std::vector<size_t>& output_shape,
+                          std::set<size_t> operated_axes,
+                          struct CollapsedDims& cdims,
+                          bool collapse_operated_axes)
+{
+    size_t collapse_size = 1;
+    bool operated_axes_run = false;
+    bool collapsing = false;
+    for (int output_idx = static_cast<int>(output_shape.size()) - 1; output_idx >= 0; output_idx--)
```

Comment:
I wonder if it might be a bit simpler to start with `output_shape().size() - 2`
```
operated_axes_run = operated_axes.count(output_shape().size() -1)
collapse_size = output_shape.at(output_shape().size()-1)
```
This way we should have no need for `collapsing`?